### PR TITLE
Optimize sequence type usage on CUDA [2/n]

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: 

1) Register support for sequences in the CUDA kernels of `Loop` and `If`. 

2) An optimization is added for when loop carried dependencies are sequences and iteration count is > 0. In that case, we can directly move these loop carried dependencies (these are subgraph outputs) directly into the Loop's output after all the iterations - there is no need to make a copy of it. We only need to make a copy when iteration count == 0 as we can't move the Loop's inputs to the outputs.

3) We use `GPUDataTransfer` (the `IDataTransfer` abstraction) to handle cross device copies in the Loop's implementation. This is okay only as long as the copy stream being used in the data transfer class is the same as the compute stream. Otherwise, we would need explicit syncs to avoid data races. As such there is no real use-case for using a dedicated copy stream (that is different from the compute stream) as there are larger negative implications. So, as a safety check, we enforce that the copy stream is the same as the compute stream in CUDA Loop. 

No new unit tests required. Unit tests for sequences were already added when the type was supported in the CPU kernels of Loop and If.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/controlflow/loop.cc [hunk 1]
```diff
@@ -83,6 +82,21 @@ static Status ConcatenateGpuOutput(void* stream, std::vector<OrtValue>& per_iter
...
+  // the compute stream to avoid data races. At the very least, we need to expose an interface in IDataTransfer
+  // to use a caller provided stream for Loop to provide for the GPUDataTransfer instance to use.
+  // Currently, using a dedicated copy stream has larger negative implications (see comment in GPUDataTransfer's
+  // constructor implementation), and so it is not in a useable state. When it becomes useable again,
+  // we will re-visit this limitation in Loop.
+  bool do_copy_on_default_stream = static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider())->DoCopyOnDefaultStream();
+  ORT_ENFORCE(do_copy_on_default_stream,
+              "Using Loop operator on CUDA while using a dedicated stream for copying "
+              "(a stream that is different than the compute stream) is currently not supported");
   SetConcatOutputFunc(ConcatenateGpuOutput);
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/controlflow/loop.cc:100]: Is it guaranteed this value will not change after the kernels are created in session state finalization?

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 2]
```diff
@@ -492,39 +492,54 @@ Status LoopImpl::Execute(const FeedsFetchesManager& ffm) {
...
-        tensors.push_back(std::move(tmp));
+    } else if (input.IsTensorSequence()) {
+      if (iter_num_value != 0) {
+        // We can move the subgraph outputs directly into the Loop's outputs.
+        TensorSeq* output = context_.Output<TensorSeq>(output_idx);
+        *output = std::move(*input.GetMutable<TensorSeq>());
+      } else {
+        // We can't move the Loop's inputs directly into the Loop's outputs.
+        // We need to make a copy.
+        std::vector<Tensor> tensors;
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc:512]: nit: say why they can't be moved, not just that they can't. more helpful to the next developer that way, especially if they're new and don't have the implicit knowledge that the inputs must be treated as readonly.

File: onnxruntime/core/providers/cuda/controlflow/loop.cc (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cuda/controlflow/loop.cc:100]: Per my understanding, yes. They shouldn't be changed after the CUDA EP instance itself is created. These are values that flow through `OrtCudaProviderOptions` struct which reside in the constructed CUDA EP as `CUDAExecutionProviderInfo`...

File: onnxruntime/core/providers/cpu/controlflow/loop.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/controlflow/loop.cc:512]: done

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]

**Pull request merged**